runtime/lua/vim/lsp/actions.lua

```diff
@@ -0,0 +1,343 @@
+
+local actions = {}
+
+--- Location actions.
+-- Supports both Location and LocationLink
+actions.Location = {}
+
+--- Jump to the first location. Accepts Location and Location[]
+actions.Location.jump_first = function()
+  return function(_, method, location, _, _)
+    local success = structures.Location.jump(location)
+
+    if success == false then
+      log.info(method, 'No location found')
+    end
+  end
+end
+
+--- Jump to the first Location. If more than one Location is returned, quickfix list is populated
+actions.Location.jump_and_quickfix = function()
+  return function(_, method, result)
+    if result == nil or vim.tbl_isempty(result) then
+      log.info(method, 'No location found')
+      return nil
+    end
+
+    -- textDocument/definition can return Location or Location[]
+    -- https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocument_definition
+
+    structures.Location.jump(result)
+    if vim.tbl_islist(result) and #result > 1 then
+      structures.Location.set_qflist(result, true)
+    end
+  end
+end
+
+--- Preview a location in a floating windows
+--
+--@param args: Config table: ~
+--              lines_above: Number of lines to show above the location in the preview. Default 0
+--              lines_below: Number of lines to show below the location in the preview. Default 0
+actions.Location.preview = function(args)
+  args = args or {}
+
+  local lines_above = args.lines_above or 0
+  local lines_below = args.lines_below or 0
+
+  return function(_, method, location)
+    local success = structures.Location.preview(location, lines_above, lines_below)
+
+    if success == false then
+      log.info(method, 'No location found')
+      return false
+    end
+  end
+end
+
+
+--- Highlight a location
+--
+--@param args: Config table: ~
+--              higroup: Name of the highlight group
+--              timeout: Number of ms before highlight is cleared.
+actions.Location.highlight = function(args)
+  args = args or {}
+
+  local higroup = args.higroup or "IncSearch"
+  local timeout = args.timeout or 250
+
+  return function(_, method, location)
+    local success = structures.Location.highlight(location, higroup, timeout)
+
+    if success == false then
+      log.info(method, 'No location found')
+      return false
+    end
+  end
+end
+
+--- Set the quickfix list for Location or Location[]
+--
+--@param args: Config table: ~
+--              open_list: Whether to open list or not. Default true
+actions.Location.set_qflist = function(args)
+  args = args or {}
+
+  local open_list = true
+  if args.open_list ~= nil then
+    open_list = args.open_list
+  end
+
+  return function(_, method, location)
+    local success = structures.Location.set_qflist(location, open_list)
+
+    if success == false then
+      log.info(method, 'No locations found')
+      return
+    end
+  end
+end
+
+--- Set the loc list for Location or Location[]
+--
+--@param args: Config table: ~
```

Comment:
Should I just do these as separate params or something? It's kind of annoying to say the table every time.

Reply:
I guess they don't have to be in a table, but they seem weird without it and a lot less explicit.

Reply:
Agreed. will leave as table.